tests/Unit/DataStructures/Tensor/Expressions/Test_Contract.cpp

```diff
@@ -0,0 +1,744 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/Contract.hpp"
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/Requires.hpp"
+
+template <typename T,
+          Requires<std::is_same_v<typename T::type, DataVector>> = nullptr>
+void assign_unique_datavector_tensor_values(T& tensor) {
+  double value = 0.0;
+  for (auto index_it = tensor.begin(); index_it != tensor.end(); index_it++) {
+    for (auto vector_it = index_it->begin(); vector_it != index_it->end();
+         vector_it++) {
+      *vector_it = value;
+      value += 1.0;
+    }
+  }
+}
+
+SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.Expression.Contract",
+                  "[DataStructures][Unit]") {
```

Comment:
@nilsdeppe I really wanted to do something like this, but I couldn't think of anything straight-forward or that didn't lead to an explosion of helper functions to deal with different contraction cases due to difficulties generalize the `get` call in the for loops. For example, if we have some rank 3 tensor R_{iIj} contracted and want to check the values in the contracted tensor by doing:
```
for (size_t j = 0; j < 4; j++) {
    double expected_sum = 0.0;
    for (size_t i = 0; i < 3; i++) {
      expected_sum += Rlul.get(i, i, j);  // <- this call
    }
    CHECK(RiIj_contracted.get(j) == expected_sum);
  }
```
then in the call to `get` above, `i` and `j` need to be passed in at the right argument slots. If the tensor was instead R_{ijI}, then we would need to do `Rlul.get(i, j, i)`.

So, I figured that in order to generalize the testing, there would need to be a specialized function for handling each contraction case: rank 2 (only 1 case, contract 2 positions), rank 3 (1st and 2nd; 2nd and 3rd; 3rd and 4th), and rank 4 (1st and 2nd; 1st and 3rd; 1st and 4th; 2nd and 3rd; 2nd and 4th; 3rd and 4th; 1st and 2nd WITH 3rd and 4th; 1st and 3rd WITH 2nd and 4th; 1st and 4th WITH 2nd and 3rd). However, maybe not all of these need their own function to generalize testing - for example, maybe we don't need to test the double contraction cases with generalized functions and instead just have 1 test case for each, or something, in `Test_Contraction.cpp`. And then I *think* there would also need to be a `double` and `DataVector` version, but this maybe could be handled with `if constexpr` within each case's function.

What do you think? There is probably a better way than writing a specialized function for each, but I haven't thought of anything yet. I can think on this more. Even if we don't have a better way than what I described above, I'm happy to implement it!

Reply:
Just to keep a log of the discussion that happened outside of this thread, I misinterpreted the request to move the different test cases to their own functions (not make generalized functions for all cases). I'll move these into their own functions based on rank

Reply:
@nilsdeppe The rank cases have been grouped into functions. To do this, I pass in a size 5 `DataVector` as `used_for_size` for the `DataVector` case. Within the `test_contractions_rank#` functions, I initialize the tensors by passing in `5_st` to the `Tesnor` constructor. Is there a way to grab the size form the `used_for_size` `DataVector` and use that to initialize the `Tensor`s instead of relying on the size of the `DataVector` passed in being 5?

Reply:
Good question! I think you can just pass `used_for_size` to the tensor constructor :)